### DISCUSSION TITLE
Threads in ring-mqtt

### DISCUSSION THREAD

**Discussion by iskiselev** [tsightler/ring-mqtt#575] [Support Q&A]
I see strange issue with either HomeAssistant or ring-mqtt addon and trying to understand which part is responsible.
Calling service lock.unlock for lock, connected to ring-mqtt takes some time (few seconds). It is not a problem, but I see problem when I try to emit several command parallelly.
I have few independent automations on the same trigger:
 - unlock first_lock
 - unlock second_lock
 - turn on multiple lights (light_1, light_2, light_3)

Despite this actions happens in separate automation, only one action at a moment happens. So, with it I see (in that order with approximated time from trigger): 
- 0 sec: light_1 turned on
- 0 sec:  first_lock unlocked
- 2 sec: second_lock unlocked
- 4 sec: light_2 turned on
- 4 sec light_3 turned on

So, looks like something put events in queue and process them strictly consequence: either HomeAssistant, mqtt, ring-mqtt or ring.
With so slow processing, light_2 and light_3 context.parent_id is not set and my automation is unable to identify that light was turned on by that automation.
Any ideas what to check or if there any workarounds?

**Comment by tsightler** [comment]
You didn't say anything about the technology used by these devices, but my guess is that you are simply seeing the Z-wave serialization delay.  The Z-wave protocol itself is serialized, each command requires an acknowledgment before proceeding to the next command.  For lights this is usually pretty fast, but other devices aren't always as fast as they may require multiple round trips.  Locks, may require additional handshakes for security, and also add a delay due to the time required for them to actually actuate (they don't send back status until the lock is fully open or closed).

There's definitely nothing in ring-mqtt or mqtt protocol that would cause this, and I don't believe that HA has any such queue of events either.  I could be wrong about HA, as I'm not an expert on that and this isn't an HA support forum, but I automate a large number of devices across different protocols and I've never seen any behavior that indicates that events are only handled serially in a queue, other than when I need to issue mulitple commands to Z-wave devices in quick succession, but again, this is more about how the Z-wave protocol works than HA itself and this behavior is well known and discussed.
